core/src/test/java/mediumtest/websockets/WebSocketMediumTests.java

```diff
@@ -0,0 +1,726 @@
+      startWebSocketServer();
+      backend = newBackend()
+        .withSmartNotifications()
+        .withSonarCloudConnectionAndNotifications("connectionId", "orgKey", null)
+        .build();
+
+      backend.getConnectionService().didChangeCredentials(new DidChangeCredentialsParams("connectionId"));
+
+      await().pollDelay(Duration.ofMillis(200)).atMost(Duration.ofSeconds(2)).untilAsserted(() -> assertThat(webSocketServer.getConnections()).isEmpty());
+    }
+
+    @Test
+    void should_do_nothing_for_sonarqube() {
+      startWebSocketServer();
+      backend = newBackend()
+        .withSmartNotifications()
+        .withSonarQubeConnection("connectionId")
+        .withBoundConfigScope("configScope", "connectionId", "projectKey")
+        .build();
+
+      backend.getConnectionService().didChangeCredentials(new DidChangeCredentialsParams("connectionId"));
+
+      await().pollDelay(Duration.ofMillis(200)).atMost(Duration.ofSeconds(2)).untilAsserted(() -> assertThat(webSocketServer.getConnections()).isEmpty());
+    }
+  }
+
+  @Nested
+  class WhenConnectionAdded {
+    @Test
+    void should_subscribe_all_projects_bound_to_added_connection() {
+      startWebSocketServer();
+      backend = newBackend()
+        .withSmartNotifications()
+        .withBoundConfigScope("configScope", "connectionId", "projectKey")
+        .build();
+
+      backend.getConnectionService()
+        .didUpdateConnections(new DidUpdateConnectionsParams(emptyList(), List.of(new SonarCloudConnectionConfigurationDto("connectionId", "orgKey", false))));
+
+      await().atMost(Duration.ofSeconds(2)).untilAsserted(() -> assertThat(webSocketServer.getConnections())
+        .extracting(WebSocketConnection::isOpened, WebSocketConnection::getReceivedMessages)
+        .containsExactly(tuple(true, List.of("{\"action\":\"subscribe\",\"events\":[\"QualityGateChanged\"],\"project\":\"projectKey\"}"))));
+    }
+  }
+
+  @Nested
+  class WhenConnectionRemoved {
+    @Test
+    void should_close_connection() {
+      startWebSocketServer();
+      backend = newBackend()
+        .withSmartNotifications()
+        .withSonarCloudConnectionAndNotifications("connectionId", "orgKey", null)
+        .withBoundConfigScope("configScope", "connectionId", "projectKey")
+        .build();
+      await().atMost(Duration.ofSeconds(2)).until(() -> !webSocketServer.getConnections().isEmpty());
+
+      backend.getConnectionService().didUpdateConnections(new DidUpdateConnectionsParams(emptyList(), emptyList()));
+
+      await().atMost(Duration.ofSeconds(2)).untilAsserted(() -> assertThat(webSocketServer.getConnections())
+        .extracting(WebSocketConnection::isOpened, WebSocketConnection::getReceivedMessages)
+        .containsExactly(tuple(false, List.of("{\"action\":\"subscribe\",\"events\":[\"QualityGateChanged\"],\"project\":\"projectKey\"}"))));
+    }
+
+    @Test
+    void should_not_close_connection_if_another_sonarcloud_connection_is_active() {
```

Comment:
Does this test belong to the following nested class - `WhenConnectionUpdated`?

Reply:
No because we remove a connection in the test, so WhenConnectionRemoved is right here